### PR TITLE
Implement SignatureInfo helper and get_signer_organizational_unit_name shortcut

### PR DESCRIPTION
**Pull request by EnTeQuAk**
Test output:

```
py27 runtests: commands[2] | python /Users/ente/Projects/signing-clients/setup.py test
running test
running egg_info
writing top-level names to signing_clients.egg-info/top_level.txt
writing requirements to signing_clients.egg-info/requires.txt
writing signing_clients.egg-info/PKG-INFO
writing dependency_links to signing_clients.egg-info/dependency_links.txt
reading manifest file 'signing_clients.egg-info/SOURCES.txt'
reading manifest template 'MANIFEST.in'
writing manifest file 'signing_clients.egg-info/SOURCES.txt'
running build_ext
test_continuation (signing_clients.tests.SigningTest) ... ok
test_extractor (signing_clients.tests.SigningTest) ... ok
test_get_addon_type_from_signature_privileged_webext (signing_clients.tests.SigningTest) ... ok
test_get_signer_organizational_unit_name (signing_clients.tests.SigningTest) ... ok
test_line_too_long (signing_clients.tests.SigningTest) ... ok
test_make_signed (signing_clients.tests.SigningTest) ... ok
test_make_signed_default_sigpath (signing_clients.tests.SigningTest) ... ok
test_manifest (signing_clients.tests.SigningTest) ... ok
test_metainf_case_sensitivity (signing_clients.tests.SigningTest) ... ok
test_resigning_manifest_exclusions (signing_clients.tests.SigningTest) ... ok
test_serial_number_extraction (signing_clients.tests.SigningTest) ... ok
test_signature (signing_clients.tests.SigningTest) ... ok
test_signatures (signing_clients.tests.SigningTest) ... ok
test_signatures_omit (signing_clients.tests.SigningTest) ... ok
test_unicode (signing_clients.tests.SigningTest) ... ok
test_wrapping (signing_clients.tests.SigningTest) ... ok
test_issuer (signing_clients.tests.TestSignatureInfo) ... ok
test_issuer_serial_number (signing_clients.tests.TestSignatureInfo) ... ok
test_loading_pass_signature_info_instance (signing_clients.tests.TestSignatureInfo) ... ok
test_loading_reading_string (signing_clients.tests.TestSignatureInfo) ... ok
test_signer_certificate (signing_clients.tests.TestSignatureInfo) ... ok

----------------------------------------------------------------------
Ran 21 tests in 0.140s

OK
py35 inst-nodeps: /Users/ente/Projects/signing-clients/.tox/dist/signing_clients-1.2.zip
py35 installed: asn1crypto==0.22.0,signing-clients==1.2,six==1.10.0
py35 runtests: PYTHONHASHSEED='2621742052'
py35 runtests: commands[0] | pip install --upgrade pip setuptools wheel
Requirement already up-to-date: pip in ./.tox/py35/lib/python3.5/site-packages
Requirement already up-to-date: setuptools in ./.tox/py35/lib/python3.5/site-packages
Requirement already up-to-date: wheel in ./.tox/py35/lib/python3.5/site-packages
py35 runtests: commands[1] | pip install --use-wheel -e .
Obtaining file:///Users/ente/Projects/signing-clients
Requirement already satisfied: asn1crypto in ./.tox/py35/lib/python3.5/site-packages (from signing-clients==1.2)
Requirement already satisfied: six in ./.tox/py35/lib/python3.5/site-packages (from signing-clients==1.2)
Installing collected packages: signing-clients
  Found existing installation: signing-clients 1.2
    Not uninstalling signing-clients at /Users/ente/Projects/signing-clients, outside environment /Users/ente/Projects/signing-clients/.tox/py35/bin/..
  Running setup.py develop for signing-clients
Successfully installed signing-clients-1.2
py35 runtests: commands[2] | python /Users/ente/Projects/signing-clients/setup.py test
running test
running egg_info
writing dependency_links to signing_clients.egg-info/dependency_links.txt
writing top-level names to signing_clients.egg-info/top_level.txt
writing requirements to signing_clients.egg-info/requires.txt
writing signing_clients.egg-info/PKG-INFO
reading manifest file 'signing_clients.egg-info/SOURCES.txt'
reading manifest template 'MANIFEST.in'
writing manifest file 'signing_clients.egg-info/SOURCES.txt'
running build_ext
test_continuation (signing_clients.tests.SigningTest) ... /Users/ente/Projects/signing-clients/.tox/py35/lib/python3.5/re.py:203: FutureWarning: split() requires a non-empty pattern match.
  return _compile(pattern, flags).split(string, maxsplit)
ok
test_extractor (signing_clients.tests.SigningTest) ... ok
test_get_addon_type_from_signature_privileged_webext (signing_clients.tests.SigningTest) ... ok
test_get_signer_organizational_unit_name (signing_clients.tests.SigningTest) ... ok
test_line_too_long (signing_clients.tests.SigningTest) ... ok
test_make_signed (signing_clients.tests.SigningTest) ... ok
test_make_signed_default_sigpath (signing_clients.tests.SigningTest) ... ok
test_manifest (signing_clients.tests.SigningTest) ... ok
test_metainf_case_sensitivity (signing_clients.tests.SigningTest) ... ok
test_resigning_manifest_exclusions (signing_clients.tests.SigningTest) ... ok
test_serial_number_extraction (signing_clients.tests.SigningTest) ... ok
test_signature (signing_clients.tests.SigningTest) ... ok
test_signatures (signing_clients.tests.SigningTest) ... ok
test_signatures_omit (signing_clients.tests.SigningTest) ... ok
test_unicode (signing_clients.tests.SigningTest) ... ok
test_wrapping (signing_clients.tests.SigningTest) ... ok
test_issuer (signing_clients.tests.TestSignatureInfo) ... ok
test_issuer_serial_number (signing_clients.tests.TestSignatureInfo) ... ok
test_loading_pass_signature_info_instance (signing_clients.tests.TestSignatureInfo) ... ok
test_loading_reading_string (signing_clients.tests.TestSignatureInfo) ... ok
test_signer_certificate (signing_clients.tests.TestSignatureInfo) ... ok

----------------------------------------------------------------------
Ran 21 tests in 0.105s

OK
```

### CODE REVIEW

**Comment by EnTeQuAk** [comment]
r? @jvehent @eviljeff 

This is the signing-clients part of https://github.com/mozilla/addons-server/pull/5868 and https://github.com/mozilla/addons-server/issues/5619 and should implement a more future-proof version of getting our signature serial number, issuer and ou.

File: signing_clients/apps.py [hunk 1]
```diff
@@ -102,6 +106,21 @@ def _digest(data):
     return {'md5': md5.digest(), 'sha1': sha1.digest()}
 
 
+def get_addon_type_from_signature(pkcs7):
+    info = SignatureInfo(pkcs7)
+    ou = get_signer_organizational_unit_name(info)
+
+    if ou == constants.OU_PRODUCTION:
+        # TODO: where should the CN be taken from to identify
+        # hotfix-addons? `if endEntity.CN = "firefox-hotfix@mozilla.org"`
```
Comment (EnTeQuAk) [inline signing_clients/apps.py]: @jvehent based on https://github.com/mozilla/addons-server/issues/5619#issuecomment-309033995 - what's `endEntity` in your code snippet referencing to in a real signature?

File: signing_clients/apps.py [hunk 2]
```diff
@@ -32,6 +35,7 @@
 directory_re = re.compile(r"[\\/]$")
 
 
+
```
Comment (eviljeff) [inline signing_clients/apps.py]: extra line

File: signing_clients/constants.py [hunk 3]
```diff
@@ -0,0 +1,8 @@
+OU_PRODUCTION = 'Production'
+OU_PRIVILEGED_WEB_EXTENSION = 'Mozilla Extensions'
+OU_SYSTEM_ADDON = 'Mozilla Components'
+
+TYPE_HOTFIX = 'hotfix'
+TYPE_ADDON = 'add-on'
+TYPE_PRIVILEGED_WEB_EXTENSION = 'privileged web extension'
+TYPE_SYSTEM_ADDON = 'system add-on'
```
Comment (eviljeff) [inline signing_clients/constants.py]: I'm +1 for constants as a rule... but we're just transposing one constant for another.  If the constants are well named, do we really need to do: `if ou == constants.OU_SYSTEM_ADDON: constants.TYPE_SYSTEM_ADDON` ?

File: signing_clients/apps.py (hunk 1 above)
Comment (jvehent) [inline signing_clients/apps.py]: The `endEntity` is the certificate that matches the `issuerAndSerialNumber` value from the `SignerInfo` sequence of the signature. See https://github.com/mozilla/addons-server/pull/5868#discussion_r126283815 for the correct way to search for it.

In the listing below, `issuerAndSerialNumber` are listed at the bottom of the signature, and point to the certificate at the top.
untitled

Comment (EnTeQuAk) [inline signing_clients/apps.py]: Perfect! So it's exactly https://github.com/mozilla/signing-clients/pull/28/files#diff-0eaa4a7d5a7900fca93854da27046e92R414 just the correct wording. Thanks!

**Comment by EnTeQuAk** [comment]
Thanks for the infos, I added support for hotfix add-ons now, and cleaned up the code a bit according to comments. ready for another review :)

**Review by eviljeff — APPROVED** [review]
lgtm

File: signing_clients/apps.py [hunk 4]
```diff
@@ -102,6 +105,21 @@ def _digest(data):
...
+    if ou == constants.OU_PRODUCTION:
+        if info.signer_certificate['issuer']['common_name'] == 'firefox-hotfix@mozilla.org':
+            return 'hotfix'
+        return 'add-on'
+    elif ou == constants.OU_PRIVILEGED_WEB_EXTENSION:
+        return 'privileged web extension'
+    elif ou == constants.OU_SYSTEM_ADDON:
+        return 'system add-on'
+    return ou
```
Comment (jvehent) [inline signing_clients/apps.py]: Why use different types here? Wouldn't it be more consistent to just return the OU?

File: signing_clients/apps.py [hunk 5]
```diff
@@ -102,6 +105,21 @@ def _digest(data):
...
+def get_addon_type_from_signature(pkcs7):
+    info = SignatureInfo(pkcs7)
+    ou = get_signer_organizational_unit_name(info)
+
+    if ou == constants.OU_PRODUCTION:
+        if info.signer_certificate['issuer']['common_name'] == 'firefox-hotfix@mozilla.org':
+            return 'hotfix'
+        return 'add-on'
+    elif ou == constants.OU_PRIVILEGED_WEB_EXTENSION:
+        return 'privileged web extension'
```
Comment (jvehent) [inline signing_clients/apps.py]: During initial discussion, we made sure to stay away from the word `privileged`, which is why they are called **Mozilla** Extensions.

File: signing_clients/apps.py [hunk 6]
```diff
@@ -367,19 +385,43 @@ def make_signed(self, signature, outpath=None, sigpath=None):
...
+            self.content = cms.ContentInfo.load(pkcs7).native['content']
+
+    @property
+    def issuer_serial_number(self):
+        return self.signer_info['sid']['serial_number']
+
+    @property
+    def signer_info(self):
+        """There should be only one SignerInfo for add-ons,
+        nss doesn't support multiples"""
```
Comment (jvehent) [inline signing_clients/apps.py]: You could link to https://bugzilla.mozilla.org/show_bug.cgi?id=1357815#c4

File: signing_clients/apps.py [hunk 7]
```diff
@@ -367,19 +385,43 @@ def make_signed(self, signature, outpath=None, sigpath=None):
...
+        if isinstance(pkcs7, SignatureInfo):
+            # Allow passing around SignatureInfo objects to avoid
+            # re-reading the signature every time.
+            self.content = pkcs7.content
+        else:
+            self.content = cms.ContentInfo.load(pkcs7).native['content']
+
+    @property
+    def issuer_serial_number(self):
+        return self.signer_info['sid']['serial_number']
```
Comment (jvehent) [inline signing_clients/apps.py]: nit on naming: `issuer` may be confusing, as what you're returning is the serial number of the end entity, not the intermediate CA that issued the end entity. `signer_serial_number` may be more accurate.

File: signing_clients/tests/__init__.py [hunk 8]
```diff
@@ -254,3 +263,127 @@ def test_metainf_case_sensitivity(self):
...
+        self.assertEqual(self.info.issuer, collections.OrderedDict([
+            ('country_name', 'US'),
+            ('state_or_province_name', 'CA'),
+            ('locality_name', 'Mountain View'),
+            ('organization_name', 'Addons Test Signing'),
+            ('common_name', 'test.addons.signing.root.ca'),
+            ('email_address', 'opsec+stagerootaddons@mozilla.com')
+        ]))
+
+    def test_signer_certificate(self):
```
Comment (jvehent) [inline signing_clients/tests/__init__.py]: Would you like some sample PKCS7 signatures for each addon type?

**Review by jvehent — APPROVED** [review]
r+ with comments

File: signing_clients/apps.py (hunk 4 above)
Comment (eviljeff) [inline signing_clients/apps.py]: Yeah, that's what I meant really when I was talking about the constants before.  Having the constants it useful but we may as well return the value of the OU.  Apart from the hotfix - that's an exception - not sure whether you want to return a false OU to identify it?

Comment (EnTeQuAk) [inline signing_clients/apps.py]: I misundertood, I'll return the OU.

File: signing_clients/apps.py (hunk 7 above)
Comment (EnTeQuAk) [inline signing_clients/apps.py]: Ah, that makes more sense. Slowly getting around all the correct vocabulary :) Thanks!

File: signing_clients/tests/__init__.py (hunk 8 above)
Comment (EnTeQuAk) [inline signing_clients/tests/__init__.py]: If you have them lying around somewhere, absolutely! I filed this as a follow-up but the more test-data we can get the better.

**Comment by EnTeQuAk** [comment]
I'm merging this, implemented all the needed changes and will follow-up on more detailed tests with more test add-ons for the OU extraction later.

**Pull request merged**